lib/resolve.js

```diff
@@ -72,37 +101,62 @@ const singularize = function (matches) {
   }
 }
 
-exports.attachment = function (heroku, app, id, options = {}) {
-  const headers = options.headers || {}
+const appAttachment = function (heroku, app, id, options = {}) {
+  const headers = addActionsHeader(options.headers)
+  return heroku.post('/actions/addon-attachments/resolve', {
+    'headers': headers, 'body': {'app': app, 'addon_attachment': id, 'addon_service': options.addon_service || null}
+  }).then((attachments) => singularize(attachments))
+}
 
-  function getAttachment (id) {
-    return heroku.get(`/addon-attachments/${encodeURIComponent(id)}`, {headers})
-      .catch(function (err) { if (err.statusCode !== 404) throw err })
+exports.appAttachment = appAttachment
+
+const filter = function (attachment, app, addonService) {
+  if (attachment.app.name !== app) {
+    return false
   }
 
-  function getAppAttachment (app, id) {
-    if (!app || id.indexOf('::') !== -1) return getAttachment(id)
-    return heroku.get(`/apps/${app}/addon-attachments/${encodeURIComponent(id)}`, {headers})
+  if (addonService && attachment.addon_service.name !== addonService) {
+    return false
+  }
+
+  return true
+}
+
+exports.attachment = function (heroku, app, id, options = {}) {
+  const headers = addActionsHeader(options.headers)
+
+  function getAttachment (id) {
+    return heroku.post('/actions/addon-attachments/resolve', {
+      'headers': headers, 'body': {'app': null, 'addon_attachment': id, 'addon_service': options.addon_service || null}
+    }).then((attachments) => singularize(attachments))
       .catch(function (err) { if (err.statusCode !== 404) throw err })
   }
 
   function getAppAddonAttachment (addon, app) {
     return heroku.get(`/addons/${encodeURIComponent(addon.id)}/addon-attachments`, {headers})
       .then(function (attachments) {
-        return singularize(attachments.filter((att) => att.app.name === app))
+        return singularize(attachments.filter((att) => filter(att, app, options.addon_service)))
```

Comment:
@dickeyxxx changed to call a function which returns a function that takes the attachments array so that I could clean up the thens in https://github.com/heroku/heroku-cli-addons/pull/57/commits/004cbbe3885bf75f879a932ec6a1dd8197e4d379 let me know what you think.